File: app/src/main/java/app/we/go/emojidraw/features/practice/PracticePresenter.java
package app.we.go.emojidraw.features.practice;

import android.util.Log;

import java.util.List;

import javax.inject.Inject;

import app.we.go.emojidraw.api.EmojiDetectionProvider;
import app.we.go.emojidraw.arch.di.qualifier.PracticeSize;
import app.we.go.emojidraw.arch.di.scope.ActivityScope;
import app.we.go.emojidraw.arch.mvp.BasePresenter;
import app.we.go.emojidraw.data.EmojiToDraw;
import app.we.go.emojidraw.data.EmojiToDrawProvider;
import app.we.go.emojidraw.model.Stroke;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

@ActivityScope
public class PracticePresenter extends BasePresenter<PracticeContract.View>
        implements PracticeContract.Presenter {

    private final EmojiDetectionProvider detectionProvider;
    private final EmojiToDrawProvider emojiToDrawProvider;
    private Integer nEmojis;

    private List<EmojiToDraw> emojisToDraw;
    private int currentEmojiIndex;
    private boolean cheated;
    private boolean won;



    @Inject
    public PracticePresenter(EmojiDetectionProvider detectionProvider, EmojiToDrawProvider emojiToDrawProvider,
                             @PracticeSize Integer nEmojis) {
        this.detectionProvider = detectionProvider;
        this.emojiToDrawProvider = emojiToDrawProvider;
        this.nEmojis = nEmojis;
    }


    @Override
    protected void init() {
        super.init();
        emojisToDraw = emojiToDrawProvider.provide(nEmojis);
        updateEmojiToDraw();
    }

    private void updateEmojiToDraw() {
        if (view != null) {
            final EmojiToDraw emojiToDraw = emojisToDraw.get(currentEmojiIndex);
            view.setEmojiToDraw(emojiToDraw.emoji(), emojiToDraw.description());
        }
    }

    @Override
    public void onStrokeAdded(final List<Stroke> strokes) {
        disposables.add(
                detectionProvider.getEmojis(strokes)
                        // TODO move to provider
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(this::processEmojiGuesses,
                                this::onError)
        );

    }

    private void processEmojiGuesses(List<String> emojis) {
        if (view != null) {
            view.onGuessesReturned(emojis);
            if (!emojis.isEmpty() && currentEmojiIndex < emojisToDraw.size()) {
                final int index = emojis.indexOf(emojisToDraw.get(currentEmojiIndex).emoji());
                if (index > 0) { // if 0, user has won, so no point in showing the tooltip
                    // TODO only show the tooltip once or twice
                    view.showTooltip(index);
                } else {
                    view.hideTooltip();
                }

                String currentEmoji = emojisToDraw.get(currentEmojiIndex).emoji();

                if (emojis.get(0).equals(currentEmoji)) {
                    disposables.clear(); // dispose of all pending requests
                    if (view != null) {
                        view.onEmojiDrawnCorrectly(currentEmoji);
                    }
                    handleNextEmoji();
                }
            } else {
                view.hideTooltip();
            }
        }
    }

    private void onError(Throwable throwable) {
        Log.e("Presenter", "Error", throwable);
        if (view != null) {
            view.showErrorMessage();
        }
    }


    /**
     * @return <code>true</code> if there is a next emoji, false if we're finished
     */
    @SuppressWarnings("WeakerAccess")
    protected boolean handleNextEmoji() {
        currentEmojiIndex++;

        if (currentEmojiIndex == emojisToDraw.size()) {
            if (view != null) {
                won = true;
                if (cheated) {
                    view.onAllEmojisDrawnWithCheat();
                } else
                    view.onAllEmojisDrawn();
                }
            return false;
        } else {
            updateEmojiToDraw();
            return true;
        }
    }




    @Override
    public void onTimeExpired() {
        if (view != null && !won) {
            view.onTimeOut();
        }
    }

    @Override
    public void onSkip() {
        if (currentEmojiIndex < emojisToDraw.size() ) {
            cheated = true;
            handleNextEmoji();
        }
    }

}

File: app/src/main/java/app/we/go/emojidraw/features/practice/PracticePresenter.kt
package app.we.go.emojidraw.features.practice

import app.we.go.emojidraw.api.EmojiDetectionProvider
import app.we.go.emojidraw.arch.di.ActivityScope
import app.we.go.emojidraw.arch.mvp.BasePresenter
import app.we.go.emojidraw.data.EmojiToDraw
import app.we.go.emojidraw.data.EmojiToDrawProvider
import app.we.go.emojidraw.model.Stroke
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import javax.inject.Inject

@ActivityScope
open class PracticePresenter @Inject
constructor(private val detectionProvider: EmojiDetectionProvider,
            private val emojiToDrawProvider: EmojiToDrawProvider,
            private val nEmojis: Int) : BasePresenter<PracticeContract.View>(), PracticeContract.Presenter {

    private var emojisToDraw: List<EmojiToDraw> = emptyList()
    private var currentEmojiIndex: Int = 0
    private var hasSkipped: Boolean = false
    private var won: Boolean = false


    override fun initialize() {
        super.initialize()
        emojisToDraw = emojiToDrawProvider.provide(nEmojis)
        updateEmojiToDraw()
    }

    private fun updateEmojiToDraw() {
        val (description, emoji) = emojisToDraw[currentEmojiIndex]
        view?.setEmojiToDraw(emoji, description)
    }

    override fun onStrokeAdded(strokes: List<Stroke>) {
        disposables.add(
            detectionProvider.getEmojis(strokes)
                // TODO move to provider
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ processEmojiGuesses(it) },
                    { onError(it) })
        )

    }

    private fun processEmojiGuesses(emojis: List<String>) {
        view?.onGuessesReturned(emojis)
        if (!emojis.isEmpty() && currentEmojiIndex < emojisToDraw.size) {
            val index = emojis.indexOf(emojisToDraw[currentEmojiIndex].emoji)
            if (index > 0) { // if 0, user has won, so no point in showing the tooltip
                // TODO only show the tooltip once or twice
                view?.showTooltip(index)
            } else {
                view?.hideTooltip()
            }

            val currentEmoji = emojisToDraw[currentEmojiIndex].emoji

            if (emojis[0] == currentEmoji) {
                disposables.clear() // dispose of all pending requests
                view?.onEmojiDrawnCorrectly(currentEmoji)
                handleNextEmoji()
            }
        } else {
            view?.hideTooltip()
        }
    }

    private fun onError(throwable: Throwable) {
        Timber.e(throwable)
        view?.showErrorMessage()
    }


    /**
     * @return `true` if there is a next emoji, false if we're finished
     */
    open fun handleNextEmoji(): Boolean {
        currentEmojiIndex++

        return if (currentEmojiIndex == emojisToDraw.size) {
            won = true
            if (hasSkipped) {
                view?.onAllEmojisDrawnWithCheat()
            } else
                view?.onAllEmojisDrawn()
            false
        } else {
            updateEmojiToDraw()
            true
        }
    }


    override fun onTimeExpired() {
        if (!won) {
            view?.onTimeOut()
        }
    }

    override fun onSkip() {
        if (currentEmojiIndex < emojisToDraw.size) {
            hasSkipped = true
            handleNextEmoji()
        }
    }

}
